text/0477-policy-validation.md

```diff
@@ -0,0 +1,561 @@
+validation is applied. The goal of CDK Policy Validation is to minimize the amount of set up needed during development, and make it as easy
+as possible. For example, a typical workflow could be:
+
+> A note on the term "deployment gate" above. It is perfectly acceptable to have
+> policy at synth run as part of you CI or CD pipeline and prevent deployments.
+> What we want to ensure is that users do not think of that "deployment gate" as
+> a way to _ensure_ that their infrastructure deployment is compliant. It should
+> instead be thought of as a way to "fail fast" and report issues before
+> deployment.
+
+* Developer writes a CDK application without the correct validations config.
+* Developer deploys non-compliant stacks.
+* Deployment guardrails catch these violations, and instruct the developer to add a validations property to their
+  application.
+* Developer adds the validations property, and avoids these violations going forward.
+
+#### Alternative
+
+```ts
+const app = new App();
+Policy.of(app).add(new CfnGuardValidator());
+```
+
+When you synthesize the CDK app the validator plugins will be called and the results will be printed.
+
+```text
+$ cdk synth
+
+Validation Report (CfnGuardValidator)
+-------------------------------------
+
+(Summary)
+
+╔═══════════╤════════════════════════╗
+║ Status    │ failure                ║
+╟───────────┼────────────────────────╢
+║ Plugin    │ CfnGuardValidator      ║
+╚═══════════╧════════════════════════╝
+
+(Violations)
+
+Ensure S3 Buckets are encrypted with a KMS CMK (2 occurrences)
+
+  Occurrences:
+  
+    - Construct Path: MyStack/MyCustomL3Construct/Bucket
+    - Stack Template Path: ./cdk.out/MyStack.template.json
+    - Creation Stack:
+        └──  Bucket (MyStack/MyCustomL3Construct/Bucket)
+             │ Library: aws-cdk-lib/aws-s3.Bucket
+             │ Library Version: 2.50.0
+             │ Location: new Bucket (/home/hallcor/tmp/cdk-tmp-app/node_modules/aws-cdk-lib/aws-s3/lib/bucket.js:1:12889)
+             └──  MyCustomL3Construct (MyStack/MyCustomL3Construct)
+                  │ Library: N/A - (Local Construct)
+                  │ Library Version: N/A
+                  │ Location: new MyCustomL3Construct (/home/hallcor/tmp/cdk-tmp-app/src/custom-construct.ts:9:20)
+                  └──  MyStack (MyStack)
+                       │ Library: aws-cdk-lib.Stack
+                       │ Library Version: 2.50.0
+                       │ Location: new MyStack (/home/hallcor/tmp/cdk-tmp-app/src/main.ts:9:20)
+    - Resource Name: my-bucket
+    - Locations:
+      > BucketEncryption/ServerSideEncryptionConfiguration/0/ServerSideEncryptionByDefault/SSEAlgorithm
+
+  Recommendation: Missing value for key `SSEAlgorithm` - must specify `aws:kms`
+  How to fix:
+    > Using override `app.findChild('my-bucket').addPropertyOverride('SSEAlgorithm', 'aws:kms');`
+    > Add to construct properties for `cdk-app/MyStack/Bucket`
+      `encryption: BucketEncryption.KMS`
```

Comment:
I like this, but this needs be propped further up front that there will be hints attached for fixing violations, and that's one of the value adds of this feature!